File: Compose-Castle-App/app/src/main/java/com/example/castleapp/model/Warrior.kt
package com.example.castleapp.model

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.delay

abstract class Warrior {
    var location by mutableFloatStateOf(0f)
    abstract val delayTime: Long

    abstract suspend fun moveToCastle()
    abstract fun returnToHome()
}

class Knight(override val delayTime: Long) : Warrior() {

    override suspend fun moveToCastle() {

        while (location in 0f..99f) {
            delay(delayTime)
            location += 3
        }
    }

    override fun returnToHome() {
        location = 0f
    }
}

class Cavalry(override val delayTime: Long) : Warrior() {
    override suspend fun moveToCastle() {
        while (location in 0f..99f) {
            delay(delayTime)
            location += 5
        }
    }

    override fun returnToHome() {
        location = 0f
    }
}

File: SwiftUI-Castle-App/CastleApp/Model/Warrior.swift
//
//  Warrior.swift
//  CastleApp
//
//  Created by Jaehwa Noh on 12/29/23.
//

protocol Warrior {
    var location: Int { get set }
    var delayTime: Int { get }
    
    func moveToCastle()
    func returnToHome()
}


class Knight: Warrior {
    var location: Int
    var delayTime: Int
    
    init(delayTime: Int) {
        self.location = 0
        self.delayTime = delayTime
    }
    
    func moveToCastle() {
        while location < 100 {
            location += 3
        }
    }
    
    func returnToHome() {
        self.location = 0
    }
}

class Cavalry: Warrior {
    var location: Int
    var delayTime: Int
    
    init(delayTime: Int) {
        self.location = 0
        self.delayTime = delayTime
    }
    
    func moveToCastle() {
        while location < 100 {
            location += 5
        }
    }
    
    func returnToHome() {
        self.location = 0
    }
}
